File: backend/app/schemas/simulator.py
from __future__ import annotations
import uuid

from pydantic import BaseModel


class SimulatorConfigOut(BaseModel):
    id: uuid.UUID
    wood_type: str
    price_per_unit: float
    unit: str
    transport_cost_per_km: float
    min_transport_cost: float
    description: str | None

    model_config = {"from_attributes": True}


class SimulatorConfigCreate(BaseModel):
    wood_type: str
    price_per_unit: float
    unit: str = "m3"
    transport_cost_per_km: float = 0
    min_transport_cost: float = 0
    description: str | None = None


class SimulatorRequest(BaseModel):
    wood_type: str
    quantity: float
    distance_km: float = 0
    include_transport: bool = False


class SimulatorResult(BaseModel):
    wood_type: str
    quantity: float
    unit: str
    price_per_unit: float
    wood_cost: float
    transport_cost: float
    total: float

File: apps/web/src/api/simulator.ts
import { supabase } from "../lib/supabase";
import type { SimulatorConfig } from "../types";

export const simulatorApi = {
  configs: async () => {
    const { data } = await supabase.from("simulator_configs").select("*").order("wood_type");
    return { data: (data ?? []) as SimulatorConfig[] };
  },

  calculate: async (payload: { wood_type: string; quantity: number; distance_km: number; include_transport: boolean }) => {
    const { data: configs } = await supabase.from("simulator_configs").select("*").eq("wood_type", payload.wood_type).single();
    if (!configs) throw new Error("Tipo de madeira não encontrado");
    const wood_cost = Number(configs.price_per_unit) * payload.quantity;
    let transport_cost = 0;
    if (payload.include_transport && payload.distance_km > 0) {
      transport_cost = Math.max(Number(configs.min_transport_cost), payload.distance_km * Number(configs.transport_cost_per_km));
    }
    return { data: { wood_cost, transport_cost, total: wood_cost + transport_cost } };
  },
};
